src/core/thread/dua_manager.cpp

```diff
@@ -184,6 +241,501 @@ otError DuaManager::Store(void)
+        }
+#endif
+
+#if OPENTHREAD_CONFIG_TMF_PROXY_DUA_ENABLE
+        mChildDuaRegisteredMask.ClearAll();
+#endif
+        attempt = true;
+    }
+
+    if (attempt)
+    {
+        mRegistrationTask.Post();
+    }
+
+    ScheduleTimer();
+}
+
+void DuaManager::ScheduleTimer(void)
+{
+    if (mDelay.mValue == 0)
+    {
+        mTimer.Stop();
+    }
+    else if (!mTimer.IsRunning())
+    {
+        mTimer.Start(kStateUpdatePeriod);
+    }
+}
+
+void DuaManager::PerformNextRegistration(void)
+{
+    otError          error   = OT_ERROR_NONE;
+    Mle::MleRouter & mle     = Get<Mle::MleRouter>();
+    Coap::Message *  message = nullptr;
+    Ip6::MessageInfo messageInfo;
+    Ip6::Address     dua;
+
+    VerifyOrExit(mle.IsAttached(), error = OT_ERROR_INVALID_STATE);
+    VerifyOrExit(Get<BackboneRouter::Leader>().HasPrimary(), error = OT_ERROR_INVALID_STATE);
+
+    // Only allow one outgoing DUA.req
+    VerifyOrExit(!mIsDuaPending, error = OT_ERROR_BUSY);
+
+    // Only send DUA.req when necessary
+#if OPENTHREAD_CONFIG_DUA_ENABLE
+#if OPENTHREAD_CONFIG_TMF_PROXY_DUA_ENABLE
+    VerifyOrExit(mle.IsRouterOrLeader() || !mle.IsExpectedToBecomeRouter(), error = OT_ERROR_INVALID_STATE);
+    VerifyOrExit((mDuaState == DuaState::kToRegister && mDelay.mFields.mRegistrationDelay == 0) ||
+                     mChildDuaMask != mChildDuaRegisteredMask,
+                 error = OT_ERROR_NOT_FOUND);
+#else
+    VerifyOrExit(mDuaState == DuaState::kToRegister && mDelay.mFields.mRegistrationDelay == 0,
+                 error = OT_ERROR_NOT_FOUND);
+#endif // OPENTHREAD_CONFIG_TMF_PROXY_DUA_ENABLE
+
+    VerifyOrExit(mle.IsFullThreadDevice() || mle.GetParent().IsThreadVersion1p1(), error = OT_ERROR_INVALID_STATE);
+#endif // OPENTHREAD_CONFIG_DUA_ENABLE
+
+    // Prepare DUA.req
+    VerifyOrExit((message = Get<Coap::Coap>().NewPriorityMessage()) != nullptr, error = OT_ERROR_NO_BUFS);
+
+    SuccessOrExit(error =
+                      message->Init(OT_COAP_TYPE_CONFIRMABLE, OT_COAP_CODE_POST, OT_URI_PATH_DUA_REGISTRATION_REQUEST));
+    SuccessOrExit(error = message->SetPayloadMarker());
+
+#if OPENTHREAD_CONFIG_DUA_ENABLE
+    if (mDuaState == DuaState::kToRegister && mDelay.mFields.mRegistrationDelay == 0)
+    {
+        dua = GetDomainUnicastAddress();
+        SuccessOrExit(error = Tlv::AppendTlv(*message, ThreadTlv::kTarget, &dua, sizeof(dua)));
+        SuccessOrExit(error = Tlv::AppendTlv(*message, ThreadTlv::kMeshLocalEid, &mle.GetMeshLocal64().GetIid(),
+                                             sizeof(Ip6::InterfaceIdentifier)));
+        mDuaState             = DuaState::kRegistering;
+        mLastRegistrationTime = TimerMilli::GetNow();
+    }
+    else
+#endif // OPENTHREAD_CONFIG_DUA_ENABLE
+    {
+#if OPENTHREAD_CONFIG_TMF_PROXY_DUA_ENABLE
+        uint32_t            lastTransactionTime;
+        const Ip6::Address *duaPtr = nullptr;
+        Child *             child  = nullptr;
+
+        if (!mRegisterCurrentChildIndex)
```

Comment:
`mRegisterCurrentChildIndex` is set to True when `kDuaReRegister` is received from PBBR. 
However, the Child might update its addresses after `kDuaReRegister` is received but before next Dua registration. 

For example, after `kDuaReRegister` is received and `mRegisterCurrentChildIndex` set to true, the Child could remove its DUA and update new addresses. So, when the Parent tries to reregister for this Child, the Child does not have any Dua, causing this function to exit with `error = OT_ERROR_ABORT `. In such case, other Children might have Dua to register but couldn't get the chance. 

Thus I have this suggestion:
```suggestion
        if (!mRegisterCurrentChildIndex && IsChildHasDuaToRegister(mChildIndexDuaRegistering))
```
Where `IsChildHasDuaToRegister` is to be implemented.

Reply:
When DUA changes, the removal of old DUA and addition of new DUA is in direct function call  `MleRouter::UpdateChildAddresses()` --->`DuaManager::UpdateChildDomainUnicastAddress()`
I don't think this process will be interrupted. That is, when reregistering, either register the old one, or the new one, both seems fine to me.

Reply:
It seems fine since a device should never remove its DUA when there is a Domain Prefix in the network.

Reply:
I am thinking about the case when a Child is removed after `mRegisterCurrentChildIndex` and `mChildIndexDuaRegistering` is set for it. So, when the parent runs `PerformNextRegistration`, the Child pointed to by `mChildIndexDuaRegistering` is actually not a valid Child that has DUA to register. 

What about we reset `mRegisterCurrentChildIndex` and `mChildIndexDuaRegistering` if when corresponding Child is removed?

I think we should be more strict by asserting that if `mRegisterCurrentChildIndex == true`, the Child `mChildIndexDuaRegistering` should exist and has Dua to register. 

Thoughts? @librasungirl

Reply:
A good corner case!

I think the suggested assertion should be subset of below assertion

https://github.com/openthread/openthread/blob/4da90a3acbf0b066b43a02aa2cfa6afdf1ad4ec8/src/core/thread/dua_manager.cpp#L476

Added a new patch to reset `mRegisterCurrentChildIndex` properly when removing the child or the child's DUA. Please take a look.